### ISSUE TITLE
Many tests broken in Plone with robotsuite 2.1.1

### ISSUE THREAD

**Issue by mauritsvanrees** [collective/robotsuite#19]
See for example [this job](https://jenkins.plone.org/job/plone-5.2-python-3.7-robot-chrome/485/robot/).
I suspected my `robot_retry` change, but that is not yet used.
So I guess it is the changes from PR #16, adding support for robotframework 3.2.

Would Plone just need to update some version pins and then it will be fine?
Or would we need to change tests?

For now I have reverted the version pin change in Plone coredev 5.2 and 6.0 back to 2.0.0.

Locally with 2.1.1 in coredev 5.2 (with Python 2.7, but should not matter), you can see similar failures by trying just one test:

```
$ bin/test --all -s plone.app.multilingual -m test_robot -t 'As a site owner I can set field language independent'
...
==============================================================================
Robot                                                                         
==============================================================================
Robot.Test Schemaeditor                                                       
==============================================================================
Scenario: As a site owner I can see language independent field set... | PASS |
------------------------------------------------------------------------------
Scenario: As a site owner I can set field language independent        | FAIL |
The id "Custom" is invalid - it is already in use.
------------------------------------------------------------------------------
Scenario: As a site owner I can unset field from being language in... | FAIL |
The id "Custom" is invalid - it is already in use.
------------------------------------------------------------------------------
Robot.Test Schemaeditor                                               | FAIL |
3 critical tests, 1 passed, 2 failed
3 tests total, 1 passed, 2 failed
==============================================================================
Robot                                                                 | FAIL |
3 critical tests, 1 passed, 2 failed
3 tests total, 1 passed, 2 failed
==============================================================================
Output:  /Users/maurits/community/plone-coredev/5.2/parts/test/robot/test_schemaeditor/Scenario_As_a_site_owner_I_can_set_field_language_independent/output.xml



Failure in test Scenario As a site owner I can set field language independent (robot/test_schemaeditor.robot)
Traceback (most recent call last):
  File "/Users/maurits/.pyenv/versions/2.7.17/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/Users/maurits/community/plone-coredev/5.2/src/robotsuite/src/robotsuite/__init__.py", line 506, in runTest
    assert last_status == 'PASS', last_message
AssertionError: The id "Custom" is invalid - it is already in use.

  Ran 1 tests with 1 failures, 0 errors, 0 skipped in 14.020 seconds.
```

Weird is also that I pass `-1` to stop at the first failure, but it does not seem to listen to this.  With my retry branch, this works fine.

Should we just use newer robotframework versions, do you have a list or can you make a PR for coredev?

**Comment by datakurre** [comment]
@mauritsvanrees I’ll revert RF 3.2 related changes ASAP, because 3.2 is not yet out and figure out the issue later. Codepath for 3.1 should not have been changed, so this is a bug.

**Comment by datakurre** [comment]
@mauritsvanrees I'll released 2.0.0 https://pypi.org/project/robotsuite/2.2.0/ with RF 3.2 support related lines reverted, so all changes can be attributed to retry count option.

That said, from the the console output you attached, that exact error looks exactly what I warned about at https://github.com/collective/robotsuite/issues/17 : the robot test is retried fine, but because it is retried in too low level, test layer testTeardown/testSetup is not run and DB changes in the failed test run are not reverted :thinking:

**Comment by datakurre** [comment]
Here are the remaining changes between 2.0.0 and 2.2.1 (I released 2.2.1 due to forgetting to update changelog :see_no_evil):

```diff
diff --git a/src/robotsuite/__init__.py b/src/robotsuite/__init__.py
index 42d6c17..79ff465 100644
--- a/src/robotsuite/__init__.py
+++ b/src/robotsuite/__init__.py
@@ -244,7 +244,7 @@ class RobotTestCase(unittest.TestCase):
     def __init__(self, filename, module_relative=True, package=None,
                  source=None, name=None, tags=None, variables=None,
                  outputdir=None, setUp=None, tearDown=None, critical=None,
-                 noncritical=None, **kw):
+                 noncritical=None, retry_count=0, **kw):
         unittest.TestCase.__init__(self)
 
         filename = doctest._module_relative_path(package, filename)
@@ -297,6 +297,10 @@ class RobotTestCase(unittest.TestCase):
         self._critical = critical or []
         self._noncritical = noncritical or []
 
+        # Set number of retries that we are allowed to do on failure.
+        # Default is zero.
+        self.retry_count = retry_count
+
         # Set test fixture setup and teardown methods when given
         if setUp:
             setattr(self, 'setUp', types.MethodType(setUp, self))
@@ -344,7 +348,7 @@ class RobotTestCase(unittest.TestCase):
             writer.write_results(settings.get_rebot_settings())
         return rc
 
-    def runTest(self):
+    def _get_output_from_test_run(self):
         # Create StringIO to capture stdout into
         stdout = StringIO()
 
@@ -368,10 +372,23 @@ class RobotTestCase(unittest.TestCase):
         }
         self._runTest(self._robot_suite, **options)
         stdout.seek(0)
+        return stdout
+
+    def runTest(self):
+        # Try the test a few times.  Default: one.
+        tries = self.retry_count + 1
+        for _try in range(tries):
+            if _try:
+                print('Retrying test "{}" after failure. Retry {}.'.format(self, _try))
+            stdout = self._get_output_from_test_run()
+            # Dump stdout on test failure or error
+            if last_status != 'PASS':
+                print('\n%s' % stdout.read())
+            else:
+                if _try:
+                    print('Test "{}" passed successfully after retry {}.'.format(self, _try))
+                break
 
-        # Dump stdout on test failure or error
-        if last_status != 'PASS':
-            print('\n%s' % stdout.read())
 
         # XXX: Up to this point, everything was easy. Unfortunately, now we
         # must merge all the separate test reports into a one big summary and
@@ -386,7 +403,7 @@ class RobotTestCase(unittest.TestCase):
         # Copy screenshots in to the current working directory
         dirname = os.path.dirname(current_data_source)
         prefix = 'robot_%s_' % dirname.replace(os.path.sep, '_')
-        screenshots = re.findall('src="([^"]+\.png)"', data)
+        screenshots = re.findall(r'src="([^"]+\.png)"', data)
         for filename in screenshots:
             path = os.path.join(dirname, filename)
             if os.path.isfile(path):
@@ -395,7 +412,7 @@ class RobotTestCase(unittest.TestCase):
                     .replace(os.pardir, '')
                 shutil.copyfile(path, "%s%s" % (prefix, copy_filename))
         # Fix 'a' and 'img' tags to target the copied versions
-        data = re.sub('(href|src)="([^"]+\.png)"',
+        data = re.sub(r'(href|src)="([^"]+\.png)"',
                       '\\1="%s\\2"' % prefix, data)
 
         # Try to merge the second 'output.xml' into the first one or into the
@@ -467,6 +484,15 @@ def RobotTestSuite(*paths, **kw):
             suite.level = int(os.environ.get('ROBOTSUITE_LEVEL', 1))
         except ValueError:
             pass
+    if 'retry_count' in kw:
+        retry_count = kw.pop('retry_count')
+    else:
+        retry_count = 0
+        if 'ROBOTSUITE_RETRY_COUNT' in os.environ:
+            try:
+                retry_count = int(os.environ.get('ROBOTSUITE_RETRY_COUNT', 0))
+            except ValueError:
+                pass
     if kw.get('module_relative', True):
         kw['package'] = doctest._normalize_module(kw.get('package'))
 
@@ -492,6 +518,7 @@ def RobotTestSuite(*paths, **kw):
                                             variables=variables,
                                             source=child_suite.source,
                                             outputdir='/'.join(outputdir),
+                                            retry_count=retry_count,
                                             **kw))
                 outputdir.pop()
             for grandchild in getattr(child_suite, 'children', []):
diff --git a/src/robotsuite/tests/test_robot.py b/src/robotsuite/tests/test_robot.py
index 14fddb3..4de6086 100644
--- a/src/robotsuite/tests/test_robot.py
+++ b/src/robotsuite/tests/test_robot.py
@@ -9,7 +9,7 @@ def test_suite():
     os.environ['ROBOT_MYVAR2'] = 'ÅÄÖ'
     suite = unittest.TestSuite()
     suite.addTests([
-        robotsuite.RobotTestSuite('test_example.robot'),
+        robotsuite.RobotTestSuite('test_example.robot', retry_count=3),
         robotsuite.RobotTestSuite('test_variables.robot'),
         robotsuite.RobotTestSuite('test_setups')
     ])
```

**Comment by datakurre** [comment]
@mauritsvanrees I apologize. The console log actually shows the issues. There was regression in RF 32 support for RF < 32 causing multiple tests to be run together without isolation. RF 2.2.1 should work. And I'll figure a test for this before merging RF 32 support back again.

**Comment by mauritsvanrees** [comment]
Perfect, thanks.

**Comment by mauritsvanrees** [comment]
Confirmed fixed in `robotsuite` 2.3.0, which is now used in the Plone coredev buildout on Plone 6. Thanks!
